ambari-server/src/main/java/org/apache/ambari/server/utils/SecretReference.java

```diff
@@ -172,4 +231,56 @@ public static void replacePasswordsWithReferencesForCustomProperties(Map<String,
       }
     }
   }
+
+  /**
+   * Replaces all password type properties in the given {@link Configuration} object. Creates a new Configuration
+   * object instead of mutating the input configuration.
+   * @param configuration the input configuration
+   * @param passwordProperties password type properties in a multimap.
+   *                           It has {@code config-type -> [password-prop-1, password-prop-2, ...]} structure.
+   * @return a new configuration with password properties replaced
+   */
+  public static Configuration replacePasswordsInConfigurations(Configuration configuration,
+                                                                     Multimap<String, String> passwordProperties) {
+    // replace passwords in config properties
+    Map<String, Map<String, String>> replacedProperties = configuration.getProperties().entrySet().stream().map(
+      entry -> {
+        String configType = entry.getKey();
+        Map<String, String> replacedConfigProps =
+          replacePasswordsInPropertyMap(entry.getValue(), configType, passwordProperties);
+        return new SimpleEntry<>(configType, replacedConfigProps);
+      }).collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    // replace passwords in config attributes
+    Map<String, Map<String, Map<String, String>>> replacedAttributes = configuration.getAttributes().entrySet().stream().map(
+      configTypeEntry -> {
+        String configType = configTypeEntry.getKey();
+        Map<String, Map<String, String>> replacedConfigProps = configTypeEntry.getValue().entrySet().stream().collect(
+          toMap(Map.Entry::getKey, e -> replacePasswordsInPropertyMap(e.getValue(), configType, passwordProperties)));
+        return new SimpleEntry<>(configType, replacedConfigProps);
+      }).collect(toMap(Map.Entry::getKey, Map.Entry::getValue));
+
+    return new Configuration(replacedProperties, replacedAttributes);
+    }
+
+  /**
+   * Replaces all password type properties in the given property map. Creates a new map instead of mutating
+   * the input configuration.
+   * @param propertyMap the input property map
+   * @param passwordProperties password type properties in a multimap.
+   *                           It has {@code config-type -> [password-prop-1, password-prop-2, ...]} structure.
+   * @return a new property map with password properties replaced
+   */
+  public static Map<String, String> replacePasswordsInPropertyMap(Map<String, String> propertyMap,
+                                                                  String configType,
+                                                                  Multimap<String, String> passwordProperties) {
+    return propertyMap.entrySet().stream().map(entry -> {
+      String propertyType = entry.getKey();
+      String newValue = passwordProperties.get(configType).contains(propertyType) ?
```

Comment:
Minor issue:  If a configuration type is listed in a Blueprint/Cluster Creation template that is not a match with any known configuration type, will this throw a NullPointerException? 

This might be a problem in cases where the cluster creation template has an error (mis-spelled config type name) or if a custom config type is specified that the stacks do not include.  

Maybe we should catch this error here and log appropriately, if the code doesn't already filter out "unknown" config types at some other level.